notify.go

```diff
@@ -1,75 +1,93 @@
 package pubsub
 
 import (
+	"fmt"
+
+	"github.com/libp2p/go-libp2p/core/event"
 	"github.com/libp2p/go-libp2p/core/network"
 	"github.com/libp2p/go-libp2p/core/peer"
-	ma "github.com/multiformats/go-multiaddr"
+	"github.com/libp2p/go-libp2p/p2p/host/eventbus"
 )
 
-var _ network.Notifiee = (*PubSubNotif)(nil)
-
 type PubSubNotif PubSub
 
-func (p *PubSubNotif) OpenedStream(n network.Network, s network.Stream) {
-}
-
-func (p *PubSubNotif) ClosedStream(n network.Network, s network.Stream) {
-}
-
-func (p *PubSubNotif) Connected(n network.Network, c network.Conn) {
-	// ignore transient connections
-	if c.Stat().Transient {
-		return
+func (p *PubSubNotif) startMonitoring() error {
+	sub, err := p.host.EventBus().Subscribe([]interface{}{
+		new(event.EvtPeerConnectednessChanged),
+		new(event.EvtPeerProtocolsUpdated),
+	}, eventbus.Name("pubsub/peers-notify")) // @NOTE(gfanton): is the naming is correct ?
+	if err != nil {
+		return fmt.Errorf("unable to subscribe to EventBus: %w", err)
 	}
 
 	go func() {
-		p.newPeersPrioLk.RLock()
-		p.newPeersMx.Lock()
-		p.newPeersPend[c.RemotePeer()] = struct{}{}
-		p.newPeersMx.Unlock()
-		p.newPeersPrioLk.RUnlock()
+		defer sub.Close()
+
+		for {
+			var e interface{}
+			select {
+			case <-p.ctx.Done():
+				return
+			case e = <-sub.Out():
+			}
 
-		select {
-		case p.newPeers <- struct{}{}:
-		default:
+			switch evt := e.(type) {
+			case event.EvtPeerConnectednessChanged:
+				// send record to connected peer only
+				if evt.Connectedness == network.Connected {
+					go p.AddPeers(evt.Peer)
+				}
+			case event.EvtPeerProtocolsUpdated:
+				supportedProtocols := p.rt.Protocols()
+
+			protocol_loop:
+				for _, addedProtocol := range evt.Added {
+					for _, wantedProtocol := range supportedProtocols {
+						if wantedProtocol == addedProtocol {
+							go p.AddPeers(evt.Peer)
+							break protocol_loop
+						}
+					}
+				}
+			}
 		}
 	}()
-}
-
-func (p *PubSubNotif) Disconnected(n network.Network, c network.Conn) {
-}
 
-func (p *PubSubNotif) Listen(n network.Network, _ ma.Multiaddr) {
+	return nil
 }
 
-func (p *PubSubNotif) ListenClose(n network.Network, _ ma.Multiaddr) {
-}
-
-func (p *PubSubNotif) Initialize() {
-	isTransient := func(pid peer.ID) bool {
-		for _, c := range p.host.Network().ConnsToPeer(pid) {
-			if !c.Stat().Transient {
-				return false
-			}
```

Comment:
We want to keep this code around; please don't remove it and run it right after initializing the bus and before starting the monitoring goroutine.

Reply:
So the initialization code is still there (below), but I believe it is racey.

Reply:
See comments below.